Shark Cage/createACLinNewDesktop/createACLinNewDesktop/util.cpp

```diff
@@ -133,75 +133,106 @@ namespace util {
 	tokenDeconstructed.TokenPrivileges = (PTOKEN_PRIVILEGES) new BYTE[bufferSize];
 	GetTokenInformation(userToken, TokenPrivileges, (LPVOID)tokenDeconstructed.TokenPrivileges, bufferSize, &bufferSize);
 
 	bufferSize = 0;
 	GetTokenInformation(userToken, TokenOwner, NULL, 0, &bufferSize);
+	SetLastError(0);
 	tokenDeconstructed.TokenOwner = (PTOKEN_OWNER) new BYTE[bufferSize];
 	GetTokenInformation(userToken, TokenOwner, (LPVOID)tokenDeconstructed.TokenOwner, bufferSize, &bufferSize);
 
 	bufferSize = 0;
 	GetTokenInformation(userToken, TokenPrimaryGroup, NULL, 0, &bufferSize);
+	SetLastError(0);
 	tokenDeconstructed.TokenPrimaryGroup = (PTOKEN_PRIMARY_GROUP) new BYTE[bufferSize];
 	GetTokenInformation(userToken, TokenPrimaryGroup, (LPVOID)tokenDeconstructed.TokenPrimaryGroup, bufferSize, &bufferSize);
 
 	bufferSize = 0;
 	GetTokenInformation(userToken, TokenDefaultDacl, NULL, 0, &bufferSize);
+	SetLastError(0);
 	tokenDeconstructed.TokenDefaultDacl = (PTOKEN_DEFAULT_DACL) new BYTE[bufferSize];
 	GetTokenInformation(userToken, TokenDefaultDacl, (LPVOID)tokenDeconstructed.TokenDefaultDacl, bufferSize, &bufferSize);
 
 	bufferSize = 0;
 	GetTokenInformation(userToken, TokenSource, NULL, 0, &bufferSize);
+	SetLastError(0);
 	tokenDeconstructed.TokenSource = (PTOKEN_SOURCE) new BYTE[bufferSize];
 	GetTokenInformation(userToken, TokenSource, (LPVOID)tokenDeconstructed.TokenSource, bufferSize, &bufferSize);
 
-	//not sure about this section
-	//among others, there is a memory leak here
+	bufferSize = 0;
+	GetTokenInformation(userToken, TokenStatistics, NULL, 0, &bufferSize);
+	SetLastError(0);
+	PTOKEN_STATISTICS stats = (PTOKEN_STATISTICS) new BYTE[bufferSize];
+	GetTokenInformation(userToken, TokenStatistics, (LPVOID)stats, bufferSize, &bufferSize);
+	tokenDeconstructed.ExpirationTime = &stats->ExpirationTime;
+	tokenDeconstructed.AuthenticationId = &stats->AuthenticationId;
+
 	tokenDeconstructed.AccessMask = TOKEN_ALL_ACCESS;
+
 	PSECURITY_QUALITY_OF_SERVICE sqos =
-	new SECURITY_QUALITY_OF_SERVICE { sizeof sqos, SecurityImpersonation, SECURITY_STATIC_TRACKING, FALSE };
-	POBJECT_ATTRIBUTES oa = new OBJECT_ATTRIBUTES{ sizeof oa, 0, 0, 0, 0, sqos };
+		new SECURITY_QUALITY_OF_SERVICE{ sizeof(SECURITY_QUALITY_OF_SERVICE), stats->ImpersonationLevel, SECURITY_STATIC_TRACKING, FALSE };
+	POBJECT_ATTRIBUTES oa = new OBJECT_ATTRIBUTES{ sizeof(OBJECT_ATTRIBUTES), 0, 0, 0, 0, sqos };
 	tokenDeconstructed.ObjectAttributes = oa;
-	tokenDeconstructed.ExpirationTime = 0; //0=infinity - not supported yet
-
-	//this is very ugly memory leak
-	PLUID auth_luid_7 = new LUID(ANONYMOUS_LOGON_LUID); //this works for win 7 and below
-	PLUID auth_luid_8 = new LUID(LOCALSERVICE_LUID); //this works for win 8 and further
-	tokenDeconstructed.AuthenticationId  = auth_luid_8;
-
 }
 
-bool emergencyExit(LPTSTR pUser_name) {
-	NetLocalGroupDel(NULL, util::group_name);
-	free(pUser_name);
-	return false;
-}
 
 void enumerateSidsAndHashes(PTOKEN_ACCESS_INFORMATION pToken) {

```

Comment:
this function also doesn't seem to be used anywhere

Reply:
Fixed 😄